modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/wal/WalCorruptionTest.java

```diff
@@ -0,0 +1,215 @@
+
+                    return;
+                }
+                else {
+                    fail("Test failed with wrong exception: " + e);
+                }
+            }
+
+            fail("Test passed without expected exception.");
+        }
+    }
+
+    /** */
+    private WALIterator walIterator(File walPath, int pageSize) throws IgniteCheckedException {
+        IgniteWalIteratorFactory.IteratorParametersBuilder builder = new IgniteWalIteratorFactory.IteratorParametersBuilder()
+            .filesOrDirs(walPath);
+
+        return new StandaloneWalRecordsIteratorTest(
+            log,
+            prepareSharedCtx(pageSize),
+            new AsyncFileIOFactory(),
+            new IgniteWalIteratorFactory(log).resolveWalFiles(builder),
+            null,
+            new FileWALPointer(Long.MIN_VALUE, 0, 0),
+            new FileWALPointer(Long.MAX_VALUE, Integer.MAX_VALUE, 0),
+            false,
+            StandaloneWalRecordsIterator.DFLT_BUF_SIZE,
+            false
+        );
+    }
+
+    /**
+     * @return Fake shared context required for create minimal services for record reading.
+     */
+    @NotNull private GridCacheSharedContext prepareSharedCtx(int pageSize) throws IgniteCheckedException {
+        GridKernalContext kernalCtx = new StandaloneGridKernalContext(log, null, null);
+
+        IgniteCacheDatabaseSharedManager dbMgr = new IgniteCacheDatabaseSharedManager() {{
+            setPageSize(pageSize);
+        }};
+
+        return new GridCacheSharedContext<>(
+            kernalCtx, null, null, null,
+            null, null, null, dbMgr, null,
+            null, null, null, null, null,
+            null, null, null, null, null, null
+        );
+    }
+
+    /**
+     * Test WAL iterator.
+     */
+    private static class StandaloneWalRecordsIteratorTest extends StandaloneWalRecordsIterator {
+        /** */
+        StandaloneWalRecordsIteratorTest(
+            @NotNull IgniteLogger log,
+            @NotNull GridCacheSharedContext sharedCtx,
+            @NotNull FileIOFactory ioFactory,
+            @NotNull List<FileDescriptor> walFiles,
+            IgniteBiPredicate<WALRecord.RecordType, WALPointer> readTypeFilter,
+            FileWALPointer lowBound,
+            FileWALPointer highBound,
+            boolean keepBinary,
+            int initialReadBufferSize,
+            boolean strictBoundsCheck
+        ) throws IgniteCheckedException {
+            super(log, sharedCtx, ioFactory, walFiles, readTypeFilter, lowBound, highBound, keepBinary,
+                initialReadBufferSize, strictBoundsCheck);
+        }
+
+        /** {@inheritDoc} */
+        @Override protected IgniteCheckedException handleRecordException(
+            @NotNull Exception e,
+            @Nullable FileWALPointer ptr
+        ) {
+            return e instanceof IgniteCheckedException ? (IgniteCheckedException)e : null;
+        }
+    }
+}
```

Comment:
I don't know if our code style obliges every file to have newline in the end but all others do. I think it should be added